src/Avalonia.Base/Input/PointerOverPreProcessor.cs

```diff
@@ -0,0 +1,210 @@
+﻿using System;
+using Avalonia.Input.Raw;
+
+namespace Avalonia.Input
+{
+    internal class PointerOverPreProcessor : IObserver<RawInputEventArgs>
+    {
+        private IPointerDevice? _lastActivePointerDevice;
+        private (IPointer pointer, PixelPoint position)? _lastPointer;
+
+        private readonly IInputRoot _inputRoot;
+
+        public PointerOverPreProcessor(IInputRoot inputRoot)
+        {
+            _inputRoot = inputRoot ?? throw new ArgumentNullException(nameof(inputRoot));
+        }
+
+        public void OnCompleted()
+        {
+            ClearPointerOver();
+        }
+
+        public void OnError(Exception error)
+        {
+        }
+
+        public void OnNext(RawInputEventArgs value)
+        {
+            if (value is RawPointerEventArgs args
+                && args.Root == _inputRoot
+                && value.Device is IPointerDevice pointerDevice)
+            {
+                if (pointerDevice != _lastActivePointerDevice)
+                {
+                    ClearPointerOver();
+
+                    // Set last active device before processing input, because ClearPointerOver might be called and clear last device.
+                    _lastActivePointerDevice = pointerDevice;
+                }
+
+                if (args.Type is RawPointerEventType.LeaveWindow or RawPointerEventType.NonClientLeftButtonDown
+                    && _lastPointer is (var lastPointer, var lastPosition))
+                {
+                    _lastPointer = null;
+                    ClearPointerOver(lastPointer, args.Root, 0, args.Root.PointToClient(lastPosition),
+                        new PointerPointProperties(args.InputModifiers, args.Type.ToUpdateKind()),
+                        args.InputModifiers.ToKeyModifiers());
+                }
+                else if (pointerDevice.TryGetPointer(args) is IPointer pointer
+                    && pointer.Type != PointerType.Touch)
+                {
+                    var element = pointer.Captured ?? args.InputHitTestResult;
+
+                    SetPointerOver(pointer, args.Root, element, args.Timestamp, args.Position,
+                        new PointerPointProperties(args.InputModifiers, args.Type.ToUpdateKind()),
+                        args.InputModifiers.ToKeyModifiers());
+                }
+            }
+        }
+
+        public void SceneInvalidated(Rect dirtyRect)
+        {
+            // Pointer is outside of the target area
```

Comment:
Nit: I don't think this comment makes much sense anymore here?

Reply:
Hm, no, it doesn't.